### ISSUE TITLE
Doesn't return correct objects unless anonymous object is created

### ISSUE THREAD

**Issue by vz28bh** [glautrou/EfCoreTemporalTable#27]
Hi again, I am using EFCoreTemporalTable 1.0.3 with netcoreapp 3.1 and EntityFrameworkCore 3.1.17. I'm having an issue trying to get history from a temporal table joined to a non-temporal table. If I create a query like this where Records is temporal and ProgramType is not
```var h = _context.Records
    .AsTemporalAll()
    .Include("ProgramType")
    .Where(c => c.RecordId == recordId);
```
it returns the right number of rows but the same data (the latest record) for every row, basically ignoring the temporal syntax. In this example results, the timestamps are the same and so is the ProgramType related model.
```
[
  {
    "RecordId": 16,
    "ProgramTypeId": 3,
    "ProgramType": {
      "ProgramTypeId": 3,
      "Name": "D110s Pilot",
      "Description": "D110s Pilot",
      "IsActive": true
    },
    "PeriodStartTime": "2021-09-07T19:06:07.0977343",
    "PeriodEndTime": "9999-12-31T23:59:59.9999999"
  },
  {
    "RecordId": 16,
    "ProgramTypeId": 3,
    "ProgramType": {
      "ProgramTypeId": 3,
      "Name": "D110s Pilot",
      "Description": "D110s Pilot",
      "IsActive": true
    },
    "PeriodStartTime": "2021-09-07T19:06:07.0977343",
    "PeriodEndTime": "9999-12-31T23:59:59.9999999"
  },
```
However if I create an anonymous object then it works correctly
```
    var h = _context.Records
    .AsTemporalAll()
    .Include("ProgramType")
    .Where(c => c.RecordId == recordId)
    .Select(c => new { RecordID = c.RecordId, ProgramTypeId = c.ProgramTypeId, ProgramTypeName = c.ProgramType.Name });
```
and returns unique values for each row
```
[
  {
    "RecordID": 16,
    "ProgramTypeId": 3,
    "ProgramTypeName": "D110s Pilot",
    "PeriodStartTime": "2021-09-07T19:06:07.0977343"
  },
  {
    "RecordID": 16,
    "ProgramTypeId": 1,
    "ProgramTypeName": "D100s 3.2k",
    "PeriodStartTime": "1900-01-01T00:00:00"
  },
```
Is this an actual bug or just a limitation? Thank you

**Comment by glautrou** [comment]
Hi @vz28bh,
Can you please post the generated SQL queries for your two examples?

**Comment by keithnicholson** [comment]
Good day!  I am seeing the same behavior.  I'm using the sample database WideWorldImporters-RR.  Here is my code and  generated expression, which works well.  


```
    var colorsfullhistory = _context.Colors.AsTemporalAll()
                        .Include(c => c.LastEditedByNavigation)
                        .Where(c => c.ColorId == id);
```
```
exec sp_executesql N'SELECT [c].[ColorID], [c].[ColorName], [c].[LastEditedBy], [c].[ValidFrom], [c].[ValidTo], [p].[PersonID], [p].[CustomFields], [p].[EmailAddress], [p].[FaxNumber], [p].[FullName], [p].[HashedPassword], [p].[IsEmployee], [p].[IsExternalLogonProvider], [p].[IsPermittedToLogon], [p].[IsSalesperson], [p].[IsSystemUser], [p].[LastEditedBy], [p].[LogonName], [p].[OtherLanguages], [p].[PhoneNumber], [p].[Photo], [p].[PreferredName], [p].[SearchName], [p].[UserPreferences], [p].[ValidFrom], [p].[ValidTo]
FROM (
    SELECT * FROM [Warehouse].[Colors] FOR SYSTEM_TIME ALL
) AS [c]
INNER JOIN [Application].[People] AS [p] ON [c].[LastEditedBy] = [p].[PersonID]
WHERE [c].[ColorID] = @__id_1',N'@__id_1 int',@__id_1=12
```
For my POC I have three results.


```
ColorID	ColorName	LastEditedBy	ValidFrom	ValidTo	PersonID	CustomFields	EmailAddress	FaxNumber	FullName	HashedPassword	IsEmployee	IsExternalLogonProvider	IsPermittedToLogon	IsSalesperson	IsSystemUser	LastEditedBy	LogonName	OtherLanguages	PhoneNumber	Photo	PreferredName	SearchName	UserPreferences	ValidFrom	ValidTo
12	Super Gray	10	2021-05-19 13:48:19.2158399	9999-12-31 23:59:59.9999999	10	{ "OtherLanguages": ["Dutch","Finnish","Lithuanian"] ,"HireDate":"2007-11-17T00:00:00","Title":"Warehouse Supervisor"}	stellar@wideworldimporters.com	(415) 555-0103	Stella Rosenhain	0x1BA4B55887E2BDCB06087A20E1CC608ADDCA538BABEC0441D2D6704DCAFE2EA4	1	0	1	0	1	1	stellar@wideworldimporters.com	["Dutch","Finnish","Lithuanian"]	(415) 555-0102	NULL	Stella	Stella Stella Rosenhain	{"theme":"dark-hive","dateFormat":"mm/dd/yy","timeZone": "PST","table":{"pagingType":"numbers","pageLength": 10},"favoritesOnDashboard":true}	2016-05-31 23:14:00.0000000	9999-12-31 23:59:59.9999999
12	Gray	1	2013-01-01 00:00:00.0000000	2016-01-01 16:00:00.0000000	1	NULL	NULL	NULL	Data Conversion Only	NULL	0	0	0	0	0	1	NO LOGON	NULL	NULL	NULL	Data Conversion Only	Data Conversion Only Data Conversion Only	{"theme":"blitzer","dateFormat":"yy-mm-dd","timeZone": "PST","table":{"pagingType":"full_numbers","pageLength": 25},"favoritesOnDashboard":true}	2016-05-31 23:14:00.0000000	9999-12-31 23:59:59.9999999
12	Steel Gray	9	2016-01-01 16:00:00.0000000	2021-05-19 13:48:19.2158399	9	{ "OtherLanguages": [] ,"HireDate":"2007-12-07T00:00:00","Title":"General Manager"}	alicaf@wideworldimporters.com	(415) 555-0103	Alica Fatnowna	0x7DFAB08E9AC574C5B15CF19D18E5B3EB466EAC7392F4295815B08221E78EA790	1	0	1	0	1	1	alicaf@wideworldimporters.com	[]	(415) 555-0102	NULL	Alica	Alica Alica Fatnowna	{"theme":"humanity","dateFormat":"mm/dd/yy","timeZone": "PST","table":{"pagingType":"simple","pageLength": 10},"favoritesOnDashboard":true}	2016-05-31 23:14:00.0000000	9999-12-31 23:59:59.9999999
```

I note a few things: 

For just the color table portion: the ValidFrom and ValidTo are correct.
```
ColorID	ColorName	LastEditedBy	ValidFrom	ValidTo
12	Super Gray	10	2021-05-19 13:48:19.2158399	9999-12-31 23:59:59.9999999
12	Gray	1	2013-01-01 00:00:00.0000000	2016-01-01 16:00:00.0000000
12	Steel Gray	9	2016-01-01 16:00:00.0000000	2021-05-19 13:48:19.2158399
```
Then the People portion returns back all the fields including its ValidFrom and ValidTo.  The name of the returned fields match.  Maybe just an observation.

However when I expand the IQueryable in VS Locals, at the breakpoint, I see only the same data for 'Super Gray' three times.

<snip>
```
  | ◢ [0] | {TemporalBlazor.Models.Colors} | TemporalBlazor.Models.Colors
  | ColorId | 12 | int
  | ColorName | "Super Gray" | string
  | LastEditedBy | 10 | int
  | ▶ LastEditedByNavigation | {TemporalBlazor.Models.People} | TemporalBlazor.Models.People
  | ▶ StockItems | Count = 0 | System.Collections.Generic.ICollection<TemporalBlazor.Models.StockItems>   {System.Collections.Generic.HashSet<TemporalBlazor.Models.StockItems>}
  | ▶ ValidFrom | {5/19/2021 1:48:19 PM} | System.DateTime
  | ▶ ValidTo | {12/31/9999 11:59:59 PM} | System.DateTime
  | ◢ [1] | {TemporalBlazor.Models.Colors} | TemporalBlazor.Models.Colors
  | ColorId | 12 | int
  | ColorName | "Super Gray" | string
  | LastEditedBy | 10 | int
  | ▶ LastEditedByNavigation | {TemporalBlazor.Models.People} | TemporalBlazor.Models.People
  | ▶ StockItems | Count = 0 | System.Collections.Generic.ICollection<TemporalBlazor.Models.StockItems>   {System.Collections.Generic.HashSet<TemporalBlazor.Models.StockItems>}
  | ▶ ValidFrom | {5/19/2021 1:48:19 PM} | System.DateTime
  | ▶ ValidTo | {12/31/9999 11:59:59 PM} | System.DateTime
  | ◢ [2] | {TemporalBlazor.Models.Colors} | TemporalBlazor.Models.Colors
  | ColorId | 12 | int
  | ColorName | "Super Gray" | string
  | LastEditedBy | 10 | int
  | ▶ LastEditedByNavigation | {TemporalBlazor.Models.People} | TemporalBlazor.Models.People
  | ▶ StockItems | Count = 0 | System.Collections.Generic.ICollection<TemporalBlazor.Models.StockItems>   {System.Collections.Generic.HashSet<TemporalBlazor.Models.StockItems>}
  | ▶ ValidFrom | {5/19/2021 1:48:19 PM} | System.DateTime
```

**Comment by ksvLeo** [comment]
Hello from Uruguay! I'd like to start by thanking you for your work, with EFCore support for Temporal Tables not being an option until VS22 is out, your library comes in the utmost handy.

Having said that i'm having the same trouble, the number of results are correct, but from the instant that the recordsare requested from the context, like this;

![imagen](https://user-images.githubusercontent.com/89048796/139948501-24891eef-eb19-40c2-90ab-be088b5c4d74.png)

They are all filled the same like this: 

![imagen](https://user-images.githubusercontent.com/89048796/139948630-7c01ebcf-53c6-4db8-b257-cb75a0888bbd.png)

# Although, when i run the same query that looks like this:
     (I just don't really know how to get this raw text style without the title)
      Executed DbCommand (6ms) [Parameters=[@__request_TeammemberId_1='?' (DbType = Int32), @__request_ProjectId_2='?' (DbType = Int32)], CommandType='Text', CommandTimeout='30']
      SELECT [p].[Id], [p].[BillingRate], [p].[CeaseDate], [p].[CurrencyId], [p].[ProjectId], [p].[RoleId], [p].[StartDate], [p].[Status], [p].[TeamMemberId], [p].[WeeklyCapacity], [p0].[Id], [p0].[BillableType], [p0].[ClientId], [p0].[Created], [p0].[CreatedBy], [p0].[CurrencyId], [p0].[DefaultHourlyRate], [p0].[FinishDate], [p0].[FixedFee], [p0].[LastModified], [p0].[LastModifiedBy], [p0].[Name], [p0].[Notes], [p0].[ProjectStatus], [p0].[ProjectType], [p0].[ReferenceCode], [p0].[StartDate], [c].[Id], [c].[Address], [c].[Created], [c].[CreatedBy], [c].[CurrencyId], [c].[LastModified], [c].[LastModifiedBy], [c].[Name], [c].[Status], [t].[Id], [t].[CostRate], [t].[Name], [t].[Status]
      FROM (
          SELECT * FROM [ProjectTeams] FOR SYSTEM_TIME ALL
      ) AS [p]
      INNER JOIN [Projects] AS [p0] ON [p].[ProjectId] = [p0].[Id]
      INNER JOIN [Clients] AS [c] ON [p0].[ClientId] = [c].[Id]
      LEFT JOIN [TeamRoles] AS [t] ON [p].[RoleId] = [t].[Id]
      WHERE ([p].[TeamMemberId] = @__request_TeammemberId_1) AND ([p].[ProjectId] = @__request_ProjectId_2)
      ORDER BY [p0].[Name] 

I get the the expected results:

![imagen](https://user-images.githubusercontent.com/89048796/139949678-cbbec2dc-4c03-406e-b391-e8242abbf351.png)


Thought of opening a thread to get in contact with you until I saw this one, anyway, hope your day is good.

**Comment by NelsonLamprecht** [comment]
I'm pretty sure the issue is related to having the same DBContext already in memory. For us, this was manifested when we created a repository around the particular table using the same injected dbcontext. Since it is the same context, you are going to have the same entity. I'm not a fan of a repository pattern around things that are already repository patterns but my current company nearly insists on it at this point. I added as no tracking here [(TemporalExtension.cs),](https://github.com/glautrou/EfCoreTemporalTable/blob/9ecf1856b025dbe3264dba76ee718752b799f48d/src/TemporalExtensions.cs#L112) I was able to get the temporal table entities with the expected values. This makes sense so I can get a non tracked entity that I would never want to make changes on. 
Additionally, it appears the ef core 6 implementation also does the asnotracking: 
https://github.com/dotnet/efcore/blob/252ece7a6bdf14139d90525a4dd0099616a82b4c/src/EFCore.SqlServer/Extensions/SqlServerDbSetExtensions.cs
This may have been part of the behavior change:
https://docs.microsoft.com/en-us/ef/core/querying/tracking
"_Since a tracking query uses the change tracker, EF Core will do identity resolution in a tracking query. When materializing an entity, EF Core will return the same entity instance from the change tracker if it's already being tracked. If the result contains the same entity multiple times, you get back same instance for each occurrence. No-tracking queries don't use the change tracker and don't do identity resolution. So you get back a new instance of the entity even when the same entity is contained in the result multiple times. This behavior was different in versions before EF Core 3.0, see previous versions._"

I've opened a PR....
